src/cargo/ops/cargo_new.rs

```diff
@@ -797,3 +791,40 @@ fn discover_author() -> CargoResult<(String, Option<String>)> {
 
     Ok((name, email))
 }
+
+fn find_git_config() -> Option<GitConfig> {
+    let inside_tests = env::var("__CARGO_TEST_ROOT");
+
+    if let Ok(test_root) = inside_tests {
+        find_tests_git_config(test_root)
+    } else {
+        find_real_git_config()
+    }
+}
+
+fn find_tests_git_config(cargo_test_root: String) -> Option<GitConfig> {
+    // Path where 'git config --local' puts variables when run from inside a test
+    let test_git_config = PathBuf::from(cargo_test_root).join(".git").join("config");
+    let test_git_config = test_git_config.as_path();
+
+    if test_git_config.exists() {
+        GitConfig::open(test_git_config).ok()
+    } else {
+        GitConfig::open_default().ok()
+    }
+}
+
+fn find_real_git_config() -> Option<GitConfig> {
+    let current_working_directory = env::current_dir();
+
+    if current_working_directory.is_err() {
+        return GitConfig::open_default().ok();
+    }
+
+    let current_working_directory = current_working_directory.unwrap();
+
+    GitRepository::discover(current_working_directory)
+        .and_then(|repo| repo.config())
+        .or_else(|_| GitConfig::open_default())
+        .ok()
```

Comment:
Ah I see. I was trying to cram it into one chain to avoid duplicating `GitConifg::open_default()` call (is that even possible?), but your suggestion looks quite clear